src/protocols/LoRaWAN/LoRaWAN.cpp

```diff
@@ -183,402 +500,167 @@ int16_t LoRaWANNode::setBufferSession(uint8_t* persistentBuffer) {
-    
-    // there are at most 8 channel masks present
-    cmd.cid = RADIOLIB_LORAWAN_MAC_LINK_ADR;
-    for(int i = 0; i < 8; i++) {
-      cmd.len = MacTable[RADIOLIB_LORAWAN_MAC_LINK_ADR].lenDn;
-      memcpy(cmd.payload, startMACpayload + (i * cmd.len), cmd.len);
-      // there COULD, according to spec, be an all zeroes ADR command - meh
-      if(memcmp(cmd.payload, bufferZeroes, cmd.len) == 0) {
-        break;
-      }
-      cmd.repeat = (i+1);
-      (void)execMacCommand(&cmd);
-    }
+  cid = RADIOLIB_LORAWAN_MAC_DUTY_CYCLE;
+  (void)this->getMacLen(cid, &cLen, RADIOLIB_LORAWAN_DOWNLINK);
+  memcpy(cOcts, &this->bufferSession[RADIOLIB_LORAWAN_SESSION_DUTY_CYCLE], cLen);
+  (void)execMacCommand(cid, cOcts, cLen);
 
-  }
+  cid = RADIOLIB_LORAWAN_MAC_RX_PARAM_SETUP;
+  (void)this->getMacLen(cid, &cLen, RADIOLIB_LORAWAN_DOWNLINK);
+  memcpy(cOcts, &this->bufferSession[RADIOLIB_LORAWAN_SESSION_RX_PARAM_SETUP], cLen);
+  (void)execMacCommand(cid, cOcts, cLen);
 
-  cmd.cid = RADIOLIB_LORAWAN_MAC_DUTY_CYCLE;
-  cmd.len = MacTable[RADIOLIB_LORAWAN_MAC_DUTY_CYCLE].lenDn;
-  memcpy(cmd.payload, &this->bufferSession[RADIOLIB_LORAWAN_SESSION_DUTY_CYCLE], cmd.len);
-  (void)execMacCommand(&cmd);
+  cid = RADIOLIB_LORAWAN_MAC_RX_TIMING_SETUP;
+  (void)this->getMacLen(cid, &cLen, RADIOLIB_LORAWAN_DOWNLINK);
+  memcpy(cOcts, &this->bufferSession[RADIOLIB_LORAWAN_SESSION_RX_TIMING_SETUP], cLen);
+  (void)execMacCommand(cid, cOcts, cLen);
 
-  cmd.cid = RADIOLIB_LORAWAN_MAC_RX_PARAM_SETUP;
-  cmd.len = MacTable[RADIOLIB_LORAWAN_MAC_RX_PARAM_SETUP].lenDn;
-  memcpy(cmd.payload, &this->bufferSession[RADIOLIB_LORAWAN_SESSION_RX_PARAM_SETUP], cmd.len);
-  (void)execMacCommand(&cmd);
+  cid = RADIOLIB_LORAWAN_MAC_TX_PARAM_SETUP;
+  (void)this->getMacLen(cid, &cLen, RADIOLIB_LORAWAN_DOWNLINK);
+  memcpy(cOcts, &this->bufferSession[RADIOLIB_LORAWAN_SESSION_TX_PARAM_SETUP], cLen);
+  (void)execMacCommand(cid, cOcts, cLen);
 
-  cmd.cid = RADIOLIB_LORAWAN_MAC_RX_TIMING_SETUP;
-  cmd.len = MacTable[RADIOLIB_LORAWAN_MAC_RX_TIMING_SETUP].lenDn;
-  memcpy(cmd.payload, &this->bufferSession[RADIOLIB_LORAWAN_SESSION_RX_TIMING_SETUP], cmd.len);
-  (void)execMacCommand(&cmd);
+  cid = RADIOLIB_LORAWAN_MAC_ADR_PARAM_SETUP;
+  (void)this->getMacLen(cid, &cLen, RADIOLIB_LORAWAN_DOWNLINK);
+  memcpy(cOcts, &this->bufferSession[RADIOLIB_LORAWAN_SESSION_ADR_PARAM_SETUP], cLen);
+  (void)execMacCommand(cid, cOcts, cLen);
 
-  cmd.cid = RADIOLIB_LORAWAN_MAC_ADR_PARAM_SETUP;
-  cmd.len = MacTable[RADIOLIB_LORAWAN_MAC_ADR_PARAM_SETUP].lenDn;
-  memcpy(cmd.payload, &this->bufferSession[RADIOLIB_LORAWAN_SESSION_ADR_PARAM_SETUP], cmd.len);
-  (void)execMacCommand(&cmd);
+  cid = RADIOLIB_LORAWAN_MAC_REJOIN_PARAM_SETUP;
+  (void)this->getMacLen(cid, &cLen, RADIOLIB_LORAWAN_DOWNLINK);
+  memcpy(cOcts, &this->bufferSession[RADIOLIB_LORAWAN_SESSION_REJOIN_PARAM_SETUP], cLen);
+  (void)execMacCommand(cid, cOcts, cLen);
```

Comment:
My copy-paste sense is tingling here. A loop that iterates over an array of CID values would be more elegant.